packages/jsii-kernel/lib/objects.ts

```diff
@@ -79,26 +95,37 @@ export class ObjectTable {
   private objects: { [objid: string]: RegisteredObject } = { };
   private nextid = 10000;
 
+  public constructor(private readonly resolveType: (fqn: string) => spec.Type) { }
+
   /**
      * Register the given object with the given type
      *
      * Return the existing registration if available.
      */
-  public registerObject(obj: object, fqn: string): api.ObjRef {
+  public registerObject(obj: object, fqn: string, interfaces?: string[]): api.AnnotatedObjRef {
     if (fqn === undefined) {
       throw new Error('FQN cannot be undefined');
     }
 
     const existingRef = objectReference(obj);
     if (existingRef) {
+      if (interfaces) {
+        const allIfaces = new Set(interfaces);
+        for (const iface of existingRef[api.TOKEN_INTERFACES] || []) {
+          allIfaces.add(iface);
+        }
+        existingRef[api.TOKEN_INTERFACES] = this.removeRedundant(Array.from(allIfaces), fqn);
```

Comment:
What if the same anonymous object gets returned under 2 different interfaces? This code path will not be hit, right? Does that interface get 'added' to the list of interfaces for this object?

Reply:
On the other hand, `removeRedundant` feels fairly expensive, I wouldn't want to do that every time an object crosses the jsii boundary.

Reply:
This is still a WIP, but yeah, I have a local commit that also adds new interfaces (& removes redundant) when we return an object through a different "persona".

But I also worry about the intrinsic cost of doing this, so it might warrant a slightly deeper refactor to improve there...